core/internal/filetransfer/file_transfer_gcs.go

```diff
@@ -54,5 +77,228 @@ func (ft *GCSFileTransfer) Upload(task *ReferenceArtifactUploadTask) error {
 func (ft *GCSFileTransfer) Download(task *ReferenceArtifactDownloadTask) error {
 	ft.logger.Debug("GCSFileTransfer: Download: downloading file", "path", task.Path, "ref", task.Reference)
 
-	return fmt.Errorf("not implemented yet")
+	reference := task.Reference
+
+	// Parse the reference path to get the scheme, bucket, and object
+	bucketName, objectName, err := parseReference(reference)
+	if err != nil {
+		return formatDownloadError("error parsing reference", err)
+	}
+	bucket := ft.client.Bucket(bucketName)
+
+	// If not using checksum, we set the digest to the reference
+	if task.Digest == reference {
+		// Get all of the objects under the reference
+		var objectNames []string
+		query := &storage.Query{Prefix: objectName}
+		it := bucket.Objects(ft.ctx, query)
+		for {
+			objAttrs, err := it.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				return formatDownloadError("error iterating through objects", err)
+			}
+			objectNames = append(objectNames, objAttrs.Name)
+		}
+
+		// Try to download all objects under the reference
+		g := new(errgroup.Group)
+		g.SetLimit(maxWorkers)
+		for _, name := range objectNames {
+			objName := name // for closure in the goroutine
+			g.Go(func() error {
+				object, _, err := ft.getObjectAndAttrs(bucket, task, objName)
+				if err != nil {
+					if errors.Is(err, ErrObjectIsDirectory) {
+						ft.logger.Debug(
+							"GCSFileTransfer: Download: skipping reference because it seems to be a folder",
+							"reference", reference,
+							"object", objName,
+						)
+						return nil
+					}
+					return formatDownloadError(fmt.Sprintf("error getting object %s", objName), err)
+				}
+				localPath := getDownloadFilePath(objName, objectName, task.Path)
+				return ft.downloadFile(object, localPath)
+			})
+		}
+
+		// Wait for all of the go routines to complete and return an error if any errored out
+		if err := g.Wait(); err != nil {
+			return formatDownloadError("", err)
+		}
+	} else {
+		object, objAttrs, err := ft.getObjectAndAttrs(bucket, task, objectName)
+		if err != nil {
+			if errors.Is(err, ErrObjectIsDirectory) {
+				ft.logger.Debug(
+					"GCSFileTransfer: Download: skipping reference because it seems to be a folder",
+					"reference", reference,
+				)
+				return nil
+			}
+			return formatDownloadError(fmt.Sprintf("error getting object %s", objectName), err)
+		}
+		if objAttrs.Etag != task.Digest {
+			err := fmt.Errorf("digest/etag mismatch: etag %s does not match expected digest %s", objAttrs.Etag, task.Digest)
+			return formatDownloadError("", err)
+		}
+		return ft.downloadFile(object, task.Path)
```

Comment:
This looks very similar to the above. Breaking this method into two steps will help consolidate similar logic:

1. Get the list of files to download for the task.
2. Download each one.

Reply:
I'm not against this, but for context the second case will always only have 1 file to download. I did initially split this logic up into the two steps you outlined but changed it because it seemed like unnecessary overhead to set up the go routines when they wouldn't be useful. 

I don't have a strong opinion on this though, if we feel like this code clarity is more valuable then I can split this up!

Reply:
Goroutines are super lightweight; they're not like threads. It's OK to have an errgroup with a single goroutine. The only optimization that's important here is to skip the object prefix search when we expect a single file.

Check out the example high-level structure in my other comment---you can use a single `ft.downloadFiles` method that accepts a list of names to download.